cmd/echo/main.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/oklog/run"
+)
+
+func main() {
+	l := log.NewLogfmtLogger(log.NewSyncWriter(os.Stdout))
+	l = log.WithPrefix(l, "ts", log.DefaultTimestampUTC)
+
+	var g run.Group
+
+	// proxy code loop
+	{
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		g.Add(func() error {
+			return loop(l, ctx)
+		}, func(error) {
+			cancel()
+		})
+	}
+
+	// signal termination
+	{
+		sigterm := make(chan os.Signal, 1)
+		g.Add(func() error {
+			signal.Notify(sigterm, syscall.SIGINT, syscall.SIGTERM)
+			if sig, ok := <-sigterm; ok {
+				l.Log("msg", "stopping the proxy", "signal", sig.String())
+			}
+			return nil
+		}, func(error) {
+			signal.Stop(sigterm)
+			close(sigterm)
+		})
+	}
+
+	err := g.Run()
+	if err != nil {
+		l.Log("msg", "terminating after error", "err", err)
+		os.Exit(1)
+	}
+}
+
+func loop(l log.Logger, ctx context.Context) error {
+	var (
+		addr    string
+		tlsCert string
+		tlsKey  string
+	)
+
+	flag.StringVar(&addr, "listen", "127.0.0.1:853", "UDP address to listen on.")
+	flag.StringVar(&tlsCert, "cert", "cert.pem", "TLS certificate path.")
+	flag.StringVar(&tlsKey, "key", "key.pem", "TLS key path.")
+
+	flag.Parse()
+
+	cert, err := tls.LoadX509KeyPair(tlsCert, tlsKey)
+	if err != nil {
+		return fmt.Errorf("load certificate: %w", err)
+	}
+
+	tls := tls.Config{
+		Certificates: []tls.Certificate{cert},
+		NextProtos:   []string{"doq"},
+	}
+
+	listener, err := quic.ListenAddr(addr, &tls, nil)
+	if err != nil {
+		return fmt.Errorf("listen: %w", err)
+	}
+	defer listener.Close()
+
+	l.Log("msg", "listening for clients", "addr", addr)
+
+	wg := sync.WaitGroup{}
+
+	for {
+		session, err := listener.Accept(ctx)
+		if err != nil {
+			wg.Wait()
+			return fmt.Errorf("accept connection: %w", err)
+		}
+
+		l := log.With(l, "client", session.RemoteAddr())
+		wg.Add(1)
+		go func() {
+			handleClient(l, ctx, session)
+			wg.Done()
+		}()
+	}
+
+}
+
+func handleClient(l log.Logger, ctx context.Context, session quic.Connection) {
+	l.Log("msg", "session accepted")
+
+	var (
+		err error
+		wg  sync.WaitGroup = sync.WaitGroup{}
+	)
+
+	defer func() {
+		msg := ""
+		if err != nil {
+			msg = err.Error()
+		}
+		session.CloseWithError(0, msg)
+
+		l.Log("msg", "session closed")
+	}()
+
+	for {
+		stream, err := session.AcceptStream(ctx)
+		if err != nil {
+			break
+		}
+
+		l := log.With(l, "stream_id", stream.StreamID())
+		l.Log("msg", "stream accepted")
+
+		wg.Add(1)
+		go func() {
+			defer func() {
+				wg.Done()
+				l.Log("msg", "stream closed")
+			}()
+
+			if err := handleStream(l, stream); err != nil {
+				l.Log("msg", "stream failure", "err", err)
+			}
+		}()
+	}
+
+	wg.Wait()
+}
+
+func handleStream(l log.Logger, stream quic.Stream) error {
+	data := make([]byte, 2048)
+	n, err := stream.Read(data)
+	if err == io.EOF {
+		defer stream.Close()
+	} else if err != nil {
+		return fmt.Errorf("read query: %w", err)
+	}
+
+	_, err = stream.Write(data[:n])
```

Comment:
I think it would be useful if the echo server didn't blindly reflect any payload but at least checked that it has a DNS header with the QR bit cleared and also to set the QR bit in response.

Reply:
I had something like that in place at one point, but removed it - I think it is pretty useful to have even just the blind reflection, as that can be used for non-DNS-related testing.

Maybe a good compromise would be to have QR bit handling available as an option? If that would be acceptable, would you rather have it in "DNS-mode" or "non-DNS-mode" by default? I'm not really inclined either way :)

Reply:
I added a `-dns` option (usable as `-dns=true` or `-dns=false`), defaulting to `true`. This makes the server parse the DNS query, check that the QR bit is cleared, and set the QR bit for the response, as suggested. When `-dns=false`, the server just blindly reflects the payload as it was originally.